### ISSUE TITLE
We shouldn't evict images from the cache when they're on the screen

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#28250]
It just causes us to immediately refetch them.

**Comment by Hixie** [comment]
(I have a case where I have two images, both of which slightly more than the maximum cache size. Every time the screen is updated, one gets evicted and immediately refetched from the network, then when it comes in the other gets evicted and immediately refetched from the network, and this continues forever.)

**Comment by Hixie** [comment]
This seems to do the trick:

```diff
diff --git a/packages/flutter/lib/src/painting/image_cache.dart b/packages/flutter/lib/src/painting/image_cache.dart
index 8beb77b..f80b847 100644
--- a/packages/flutter/lib/src/painting/image_cache.dart
+++ b/packages/flutter/lib/src/painting/image_cache.dart
@@ -183,28 +183,35 @@ class ImageCache {
       _checkCacheSize();
     }
     if (maximumSize > 0 && maximumSizeBytes > 0) {
       _pendingImages[key] = _PendingImage(result, listener);
       result.addListener(listener);
     }
     return result;
   }
 
   // Remove images from the cache until both the length and bytes are below
-  // maximum, or the cache is empty.
+  // maximum, or the cache is empty (or contains only images that cannot be
+  // evicted).
   void _checkCacheSize() {
-    while (_currentSizeBytes > _maximumSizeBytes || _cache.length > _maximumSize) {
-      final Object key = _cache.keys.first;
-      final _CachedImage image = _cache[key];
-      _currentSizeBytes -= image.sizeBytes;
+    int index = 0;
+    final List<Object> keys = _cache.keys.toList();
+    while (index < keys.length &&
+           (_currentSizeBytes > _maximumSizeBytes || _cache.length > _maximumSize)) {
+      final Object key = keys[index];
+      final _CachedImage entry = _cache[key];
+      if (!entry.completer.preventEviction) {
+        _currentSizeBytes -= entry.sizeBytes;
         _cache.remove(key);
       }
+      index += 1;
+    }
     assert(_currentSizeBytes >= 0);
     assert(_cache.length <= maximumSize);
     assert(_currentSizeBytes <= maximumSizeBytes);
   }
 }
 
 class _CachedImage {
   _CachedImage(this.completer, this.sizeBytes);
 
   final ImageStreamCompleter completer;
diff --git a/packages/flutter/lib/src/painting/image_stream.dart b/packages/flutter/lib/src/painting/image_stream.dart
index 335f1d0..8e2c5a4 100644
--- a/packages/flutter/lib/src/painting/image_stream.dart
+++ b/packages/flutter/lib/src/painting/image_stream.dart
@@ -246,20 +246,34 @@ abstract class ImageStreamCompleter extends Diagnosticable {
   ///
   /// Typically this is used by overriding [addListener], checking if
   /// [hasListeners] is false before calling `super.addListener()`, and if so,
   /// starting whatever work is needed to determine when to call
   /// [notifyListeners]; and similarly, by overriding [removeListener], checking
   /// if [hasListeners] is false after calling `super.removeListener()`, and if
   /// so, stopping that same work.
   @protected
   bool get hasListeners => _listeners.isNotEmpty;
 
+  /// Whether the image should be pinned in the cache even if the cache is over-
+  /// subscribed.
+  ///
+  /// This returns true if the completer has any listeners, on the assumption
+  /// that images that have listeners are actively being used (e.g. displayed
+  /// on the screen).
+  ///
+  /// If an image is on the screen, then it should not be evicted from the
+  /// cache, because otherwise it would need to be refetched (possibly over the
+  /// network) as soon as the screen is updated (since the image wouldn't be
+  /// found in the cache). Even if this happens quickly, it can still cause
+  /// flickering.
+  bool get preventEviction => hasListeners;
+
   /// Adds a listener callback that is called whenever a new concrete [ImageInfo]
   /// object is available or an error is reported. If a concrete image is
   /// already available, or if an error has been already reported, this object
   /// will call the listener or error listener synchronously.
   ///
   /// If the [ImageStreamCompleter] completes multiple images over its lifetime,
   /// this listener will fire multiple times.
   ///
   /// The listener will be passed a flag indicating whether a synchronous call
   /// occurred. If the listener is added within a render object paint function,
```

cc @iskakaushik

**Comment by jonahwilliams** [comment]
We shouldn't cause images to be reloaded, but I disagree that we shouldn't be evicting them. Rather, we should be sure that the image widgets reference to the ui.Image is destroyed on eviction as well.

The cache needs to be more reliable, otherwise it will remain trivial to create scenarios where loading multiple high resolution images can crash applications. See: numerous bugs on image memory leaks

**Comment by dnfield** [comment]
This was fixed in a different way by #51249  - although that is temporarily reverted while I continue to figure out why it's breaking google tests. I also ended up tracking the images with listeners in a different bucket so we could more easily deal with clearing them without disrupting the rest of the cache.
